Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp

```diff
@@ -0,0 +1,200 @@
+
+namespace Acts {
+
+template <typename E, typename R, typename A>
+auto MultiEigenStepperLoop<E, R, A>::boundState(State& state,
+                                                const Surface& surface,
+                                                bool transportCov) const
+    -> Result<BoundState> {
+  if (numberComponents(state) == 1) {
+    return SingleStepper::boundState(state.components.front().state, surface,
+                                     transportCov);
+  } else {  // Do the combination
+    SmallVector<std::pair<double, BoundTrackParameters>> states;
+    double accumulatedPathLength = 0.0;
+    int failedBoundTransforms = 0;
+
+    for (auto i = 0ul; i < numberComponents(state); ++i) {
+      auto bs = SingleStepper::boundState(state.components[i].state, surface,
+                                          transportCov);
+
+      if (bs.ok()) {
+        states.push_back(
+            {state.components[i].weight, std::get<BoundTrackParameters>(*bs)});
+        accumulatedPathLength +=
+            std::get<double>(*bs) * state.components[i].weight;
+      } else {
+        failedBoundTransforms++;
+      }
+    }
+
+    if (failedBoundTransforms > 0) {
+      ACTS_ERROR("Multi component bound state: "
+                 << failedBoundTransforms << " of " << numberComponents(state)
+                 << " transforms failed");
+    }
+
+    if (states.size() == 0) {
+      return MultiStepperError::AllComponentsConversionToBoundFailed;
+    }
+
+    // TODO also implement a method of using the mode of the mixture
+    const auto [params, cov] = detail::combineBoundGaussianMixture(
+        states.begin(), states.end(), [&](const auto& wbs) {
+          return std::tie(wbs.first, wbs.second.parameters(),
+                          wbs.second.covariance());
+        });
+
+    return BoundState{BoundTrackParameters(surface.getSharedPtr(), params, cov),
+                      Jacobian::Zero(), accumulatedPathLength};
+  }
+}
+
+template <typename E, typename R, typename A>
+auto MultiEigenStepperLoop<E, R, A>::curvilinearState(State& state,
+                                                      bool transportCov) const
+    -> CurvilinearState {
+  if (numberComponents(state) == 1) {
+    return SingleStepper::curvilinearState(state.components.front().state,
+                                           transportCov);
+  } else {
+    Vector4 pos4 = Vector4::Zero();
+    Vector3 dir = Vector3::Zero();
+    ActsScalar qop = 0.0;
+    BoundSymMatrix cov = BoundSymMatrix::Zero();
+    ActsScalar pathLenth = 0.0;
+
+    for (auto i = 0ul; i < numberComponents(state); ++i) {
+      const auto [cp, jac, pl] = SingleStepper::curvilinearState(
+          state.components[i].state, transportCov);
+
+      pos4 += state.components[i].weight * cp.fourPosition(state.geoContext);
+      dir += state.components[i].weight * cp.unitDirection();
+      qop += state.components[i].weight * (cp.charge() / cp.absoluteMomentum());
```

Comment:
Couldn't one do the pos4, mom4 merging first, then distil `pos4, dir, q/p` out of it?

Somehow I think this would be more stable, but I can wrong.

Reply:
Rigth now  I'm not sure how one would do this with the mom4... 

But now where I look at it, this function (and some others...) outputs garbage if the charges of the components are not equal, right? I will think this through...